### PR TITLE
SSCS-10378 Adding joint party logic

### PR DESCRIPTION
**Pull request by dholaramgurjar**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/SSCS-10378

### Change description ###

 Adding joint party logic

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsMapping.java [hunk 1]
```diff
@@ -77,6 +78,18 @@ private static int updatePartyIds(Party party, Representative rep, int maxId) {
         return newMaxId;
     }
 
+    private static int updateJointPartyIds(JointParty jointParty, Representative rep, int maxId) {
+        int newMaxId = maxId;
+        newMaxId = updateEntityId(jointParty, newMaxId);
+        if (nonNull(jointParty.getAppointee())) {
+            newMaxId = updateEntityId(jointParty.getAppointee(), newMaxId);
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsMapping.java:85]: Not sure if we need this

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 2]
```diff
@@ -100,6 +100,26 @@ public static PartyDetails createDwpPartyDetails() {
 
     public static PartyDetails createJointPartyDetails(SscsCaseData caseData) {
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:101]: Is this method called from anywhere?

**Review by TeaWithLucas — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java (hunk 2 above)
Comment (dholaramgurjar) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:101]: Yes , it's called in method buildHearingPartiesDetails() in same class

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 3]
```diff
@@ -100,6 +100,26 @@ public static PartyDetails createDwpPartyDetails() {
...
+                                                                 caseData.getAppeal().getAppellant().getId()
+        ));
+        partyDetails.partyChannelSubType(getPartyChannelSubType());
+        partyDetails.organisationDetails(getOrganisationDetails(
+            caseData.getJointParty().getOrganisation(),
+            caseData.getJointParty().getHasJointParty().getValue(),
+            caseData.getJointParty().getId()
+        ));
+        partyDetails.unavailabilityDayOfWeek(getPartyUnavailabilityDayOfWeek());
+        partyDetails.unavailabilityRanges(getPartyUnavailabilityRange(caseData.getAppeal().getHearingOptions()));
```
[the remark this answers is not in the document]
Comment (dholaramgurjar) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:122]: Which field needs to be null , also buildHearingPartiesPartyDetails method has return type as list<partydetails> which different from joint party method.

**Review by MohammedHMCTS — DISMISSED** [review]

**Review by TeaWithLucas — APPROVED** [review]

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java [hunk 4]
```diff
@@ -186,14 +185,22 @@ void buildHearingPartiesDetailsOtherParties() {
     }
 
     @DisplayName("When a valid hearing wrapper with joint party given buildHearingPartiesDetails returns the correct Hearing Parties Details")
-    @ParameterizedTest
-    @EnumSource(value = YesNo.class)
+    @Test
     @NullSource
-    void buildHearingPartiesDetailsJointParty(YesNo jointParty) {
-        // TODO SSCS-10378 - Finish Test
+    void buildHearingPartiesDetailsJointParty() {
```
Comment (TeaWithLucas) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java:190]: Can you add assertions for joint party to this test?

**Review by TeaWithLucas — CHANGES_REQUESTED** [review]

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java [hunk 5]
```diff
@@ -186,14 +185,22 @@ void buildHearingPartiesDetailsOtherParties() {
     }
 
     @DisplayName("When a valid hearing wrapper with joint party given buildHearingPartiesDetails returns the correct Hearing Parties Details")
-    @ParameterizedTest
-    @EnumSource(value = YesNo.class)
```
Comment (TeaWithLucas) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java:190]: You need to cover all conditions in your tests:
https://sonarcloud.io/component_measures?metric=new_coverage&selected=SSCSHA%3Asrc%2Fmain%2Fjava%2Fuk%2Fgov%2Fhmcts%2Freform%2Fsscs%2Fhelper%2Fmapping%2FHearingsPartiesMapping.java&view=list&pullRequest=127&id=SSCSHA

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 6]
```diff
@@ -39,7 +39,7 @@ public static List<PartyDetails> buildHearingPartiesDetails(HearingWrapper wrapp
         }
 
         if (isYes(caseData.getJointParty().getHasJointParty())) {
-            partiesDetails.add(createJointPartyDetails(caseData));
+            partiesDetails.addAll(buildHearingPartiesPartyDetails(caseData.getJointParty(),null,null,null,null,appellant.getId()));
```
Comment (nickwest1) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:42]: Cleaner to add an extra buildHearingPartiesPartyDetails method without the values where nulls have been passed in.  E.g.:
public static List<PartyDetails> buildHearingPartiesPartyDetails(Party party, String appellantId) {
    return buildHearingPartiesPartyDetails(party,null,null,null,null,appellantId);
}

**Review by nickwest1 — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java (hunk 6 above)
Comment (dholaramgurjar) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java]: Done

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java (hunk 4 above)
Comment (dholaramgurjar) [reply src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java:190]: okay , added

**Review by nickwest1 — DISMISSED** [review]

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java (hunk 4 above)
Comment (dattipoeHMCTS) [reply src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java:192]: Isn't this test made redundant by the following one?

**Review by dattipoeHMCTS — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 7]
```diff
@@ -18,7 +18,7 @@
 import static uk.gov.hmcts.reform.sscs.model.single.hearing.PartyType.IND;
 import static uk.gov.hmcts.reform.sscs.model.single.hearing.PartyType.ORG;
 
-@SuppressWarnings({"PMD.UnnecessaryLocalBeforeReturn","PMD.ReturnEmptyCollectionRatherThanNull", "PMD.GodClass"})
+@SuppressWarnings({"PMD.UnnecessaryLocalBeforeReturn","PMB.LawOfDemeter","PMD.ReturnEmptyCollectionRatherThanNull", "PMD.GodClass"})
```
[the remark this answers is not in the document]
Comment (dholaramgurjar) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:21]: After removing this , build was completing successfully but not sure who added it

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsMapping.java [hunk 8]
```diff
@@ -52,6 +52,7 @@ public static void updateIds(HearingWrapper wrapper) {
         int maxId = getMaxId(caseData.getOtherParties(), appellant, appeal.getRep());
 
         maxId = updatePartyIds(appellant, appeal.getRep(), maxId);
+        maxId = updatePartyIds(appellant, null, maxId);
```
[the remark this answers is not in the document]
Comment (TeaWithLucas) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsMapping.java:55]: no, it should be caseData.getJointParty()

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 9]
```diff
@@ -58,6 +58,10 @@ public static List<PartyDetails> buildHearingPartiesDetails(HearingWrapper wrapp
         return partiesDetails;
     }
 
+    public static List<PartyDetails> buildHearingPartiesPartyDetails(Party party, String appellantId) {
+        return buildHearingPartiesPartyDetails(party,null,null,null,null,appellantId);
```
[the remark this answers is not in the document]
Comment (dholaramgurjar) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:62]: okay

**Review by dattipoeHMCTS — APPROVED** [review]

**Review by TeaWithLucas — DISMISSED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 10]
```diff
@@ -49,7 +49,7 @@ public static List<PartyDetails> buildHearingPartiesDetails(HearingWrapper wrapp
         }
 
         if (isYes(caseData.getJointParty().getHasJointParty())) {
-            partiesDetails.add(createJointPartyDetails(caseData));
+            partiesDetails.addAll(buildHearingPartiesPartyDetails(caseData.getJointParty(),appeal.getHearingOptions(),appellant.getId()));
```
Comment (nickwest1) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:52]: Just a tidiness comment for future considerations, all on one line makes it hard to read.  Prefer to layout easier to read such as:
partiesDetails.addAll(
  buildHearingPartiesPartyDetails(
       caseData.getJointParty(),
       appeal.getHearingOptions(),
       appellant.getId()));

Comment (nickwest1) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:52]: (it looks better with the right indentation!!!)

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsMappingTest.java [hunk 11]
```diff
@@ -93,6 +96,10 @@ void updateIds() {
         assertNotNull(wrapper.getCaseData().getOtherParties().get(0).getValue().getRep().getId());
 
         assertNotNull(wrapper.getCaseData().getOtherParties().get(1).getValue().getId());
+
+        assertNotNull(wrapper.getCaseData().getJointParty().getId());
+        assertNotNull(wrapper.getCaseData().getJointParty().getAppointee().getId());
```
Comment (nickwest1) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsMappingTest.java:101]: Hamcrest is better because it's easier to read and you get better error messages - assertThat(???).isNotNull();

**Review by nickwest1 — APPROVED** [review]
A couple of comments that I don't mind if you take it or leave it for this PR, but hopefully helpful tips moving forwards ;)

**Review by augustio — APPROVED** [review]
Looks good to me.

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 12]
```diff
@@ -49,7 +49,12 @@ public static List<PartyDetails> buildHearingPartiesDetails(HearingWrapper wrapp
         }
 
         if (isYes(caseData.getJointParty().getHasJointParty())) {
-            partiesDetails.add(createJointPartyDetails(caseData));
+            partiesDetails.addAll(
+                buildHearingPartiesPartyDetails(
+                    caseData.getJointParty(),
+                    appeal.getHearingOptions(),
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:55]: I don't think you shouldn't be using the appellant's hearing options for a joint party

Comment (dholaramgurjar) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:55]: Actually Unit testcases are failing for null point exception for hearingoptions

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 13]
```diff
@@ -167,11 +166,11 @@ public static Optional<String> getIndividualPreferredHearingChannel(String heari
     }
 
     public static Optional<String> getIndividualInterpreterLanguage(HearingOptions hearingOptions) {
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:168]: can you add a unit test for getIndividualInterpreterLanguage when hearing options is null

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 14]
```diff
@@ -167,11 +166,11 @@ public static Optional<String> getIndividualPreferredHearingChannel(String heari
...
-        if (isTrue(hearingOptions.wantsSignLanguageInterpreter())) {
+        if (null != hearingOptions && isTrue(hearingOptions.wantsSignLanguageInterpreter())) {
             return getSignLanguage(hearingOptions)
                 .map(SignLanguage::getHmcReference);
         }
-        if (isYes(hearingOptions.getLanguageInterpreter())) {
+        if (null != hearingOptions && isYes(hearingOptions.getLanguageInterpreter())) {
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:173]: The null check is duplicated, it would be better to check before with a quick fail

**Review by TeaWithLucas — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 15]
```diff
@@ -164,13 +167,16 @@ public static Optional<String> getIndividualPreferredHearingChannel(String heari
...
+            if (isTrue(hearingOptions.wantsSignLanguageInterpreter())) {
+                return getSignLanguage(hearingOptions)
+                    .map(SignLanguage::getHmcReference);
+            }
+            if (isYes(hearingOptions.getLanguageInterpreter())) {
+                return getInterpreterLanguage(hearingOptions)
+                    .map(InterpreterLanguage::getHmcReference);
+            }
         }
         return Optional.empty();
```
Comment (nickwest1) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:181]: Best to get out early. It also reduces nested statements.
I.e.
if(isNull(hearingOptions)) {
return Optional.empty();
{

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java [hunk 16]
```diff
@@ -193,36 +195,95 @@ void buildHearingPartiesDetailsOtherParties() {
...
-                                        .build())
-                                .build())
+                                       .id(appellantId)
+                                       .name(Name.builder()
+                                                 .title("title")
+                                                 .firstName("first")
+                                                 .lastName("last")
+                                                 .build())
+                                       .build())
                         .build())
```
Comment (nickwest1) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java:229]: Extract to it's own method and avoid duplication in lower test method

**Review by nickwest1 — CHANGES_REQUESTED** [review]

**Review by johnoneill46 — APPROVED** [review]

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java [hunk 17]
```diff
@@ -193,36 +195,93 @@ void buildHearingPartiesDetailsOtherParties() {
...
+            .build();
+        HearingWrapper wrapper = HearingWrapper.builder()
+            .caseData(caseData)
+            .caseData(caseData)
+            .build();
 
+
+
+        List<PartyDetails> partiesDetails = HearingsPartiesMapping.buildHearingPartiesDetails(wrapper);
+        Optional<String> hearingOptions = HearingsPartiesMapping.getIndividualInterpreterLanguage(null);
```
Comment (TeaWithLucas) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java:280]: Why are you putting this test here? It needs to be tested in its own unit test

**Review by TeaWithLucas — CHANGES_REQUESTED** [review]

**Review by nickwest1 — APPROVED** [review]

**Review by johnoneill46 — DISMISSED** [review]

**Review by TeaWithLucas — DISMISSED** [review]

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java [hunk 18]
```diff
@@ -193,37 +195,56 @@ void buildHearingPartiesDetailsOtherParties() {
...
+            .caseData(caseData)
+            .caseData(caseData)
+            .build();
 
         List<PartyDetails> partiesDetails = HearingsPartiesMapping.buildHearingPartiesDetails(wrapper);
-
         assertThat(partiesDetails.stream().filter(o -> appellantId.equalsIgnoreCase(o.getPartyID())).findFirst()).isPresent();
-
+        assertThat(partiesDetails.stream().anyMatch(o -> jointPartyId.equalsIgnoreCase(o.getPartyID())));
         assertThat(partiesDetails.stream().filter(o -> "DWP".equalsIgnoreCase(o.getPartyID())).findFirst()).isNotPresent();
```
Comment (TeaWithLucas) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java:237]: For future: Better to use assertThat(...).contains(...)

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 19]
```diff
@@ -182,42 +196,38 @@ public static String getIndividualFullName(Entity entity) {
     public static String getIndividualPreferredHearingChannel(String hearingType,
                                                                         HearingSubtype hearingSubtype,
                                                                         HearingOptions hearingOptions) {
-        if (eitherNull(hearingType, hearingSubtype)) {
```
Comment (HarryH96) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:185]: instead of wrapping everything below in nonNull() 

we can keep this if check, but just return null, instead of throwing an exception

Comment (osheaf01) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:185]: Reverted, keeping the check and returning null if true.

**Review by nickwest1 — DISMISSED** [review]

**Review by johnoneill46 — DISMISSED** [review]

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java [hunk 20]
```diff
@@ -464,23 +500,19 @@ void getIndividualLastName(String value, String expected) {
     void getIndividualPreferredHearingChannelNullHearingTypeTest() {
         HearingOptions hearingOptions = HearingOptions.builder().build();
 
-        assertThatExceptionOfType(
-            IllegalStateException.class).isThrownBy(() ->
-            getIndividualPreferredHearingChannel(null,
+        assertThat(getIndividualPreferredHearingChannel(null,
                 HearingSubtype.builder().build(),
-                hearingOptions));
+                hearingOptions)).isNull();
```
Comment (HarryH96) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java:505]: Sould the name for this test be updated?
@DisplayName("When hearingType not set then throw IllegalStateException")
